progress.go

```diff
@@ -0,0 +1,165 @@
+	// should start processing the next generation.
+	StartGeneration(gen time.Time) error
+
+	// EndGeneration marks all streams in the generation as "done".
+	EndGeneration(gen time.Time) error
+
+	// GetProgress retrieves information about the progress of given shard,
+	// in a given table.
+	GetProgress(gen time.Time, table string, streamID StreamID) (Progress, error)
+
+	// SaveProgress stores information about the last cdc log record which was
+	// processed successfully. If the reader is restarted, it should resume
+	// work for this shard from the last saved
+	SaveProgress(gen time.Time, table string, streamID StreamID, progress Progress) error
+}
+
+type Progress struct {
+	Time gocql.UUID
+}
+
+// NoProgressManager does not persist the progress at all. It can be used
+// when saving progress is not necessary.
+type NoProgressManager struct{}
+
+func (*NoProgressManager) GetCurrentGeneration() (time.Time, error) {
+	return time.Time{}, nil
+}
+
+func (*NoProgressManager) StartGeneration(gen time.Time) error {
+	return nil
+}
+
+func (*NoProgressManager) EndGeneration(gen time.Time) error {
+	return nil
+}
+
+func (*NoProgressManager) GetProgress(gen time.Time, table string, streamID StreamID) (Progress, error) {
+	return Progress{}, nil
+}
+
+func (*NoProgressManager) SaveProgress(gen time.Time, table string, streamID StreamID, progress Progress) error {
+	return nil
+}
+
+type TableBackedProgressManager struct {
+	session           *gocql.Session
+	progressTableName string
+
+	// TTL to use when writing progress for a shard (a week by default).
+	// TODO: maybe not? maybe we should clean up this data manually?
+	// Progress data may be large if generations are very large
+	ttl int32
+}
+
+func NewTableBackedProgressManager(session *gocql.Session, progressTableName string) (*TableBackedProgressManager, error) {
+	tbpm := &TableBackedProgressManager{
+		session:           session,
+		progressTableName: progressTableName,
+
+		ttl: 7 * 24 * 60 * 60, // 1 week
+	}
+
+	if err := tbpm.ensureTableExists(); err != nil {
+		return nil, err
+	}
+	return tbpm, nil
+}
+
+func (tbpm *TableBackedProgressManager) ensureTableExists() error {
+	return tbpm.session.Query(
+		fmt.Sprintf(
+			"CREATE TABLE IF NOT EXISTS %s "+
+				"(generation timestamp, table_name text, stream_id blob, last_timestamp timeuuid, current_generation timestamp, "+
+				"PRIMARY KEY ((generation, table_name, stream_id)))",
+			tbpm.progressTableName,
+		),
+	).Exec()
+}
+
+func (tbpm *TableBackedProgressManager) GetCurrentGeneration() (time.Time, error) {
+	var gen time.Time
+	err := tbpm.session.Query(
+		fmt.Sprintf("SELECT current_generation FROM %s WHERE generation = ? AND table_name = ? AND stream_id = ?", tbpm.progressTableName),
+		time.Time{}, "", []byte{},
+	).Scan(&gen)
+
+	if err != nil && err != gocql.ErrNotFound {
+		return time.Time{}, err
+	}
+	return gen, nil
+}
+
+func (tbpm *TableBackedProgressManager) StartGeneration(gen time.Time) error {
```

Comment:
This means that every consumer of the previous generation finished, right?
In that case we could clear the progress of the previous generation, saving space, and then the ttls should not be necessary